server/memcached/src/main/java/org/infinispan/server/memcached/MemcachedDecoder.java

```diff
@@ -1250,7 +1257,7 @@ void queueResponse(CompletionStage<Object> operationResponse) {
       all.dependsOn(operationResponse.whenComplete(this));
       responseSent = all.freeze()
             .exceptionally(CompletableFutures.toNullFunction())
-            .thenRun(this);
+            .thenRunAsync(this, decoder.responseExecutor);
```

Comment:
Technically the `responseSent` variable above doesn't need to be volatile either. Just the `reponse` and `throwable` variables as they can be set from a different thread.